python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -176,14 +176,32 @@ def have_all_requested_runs_finished(self) -> bool:
         return True
 
     def get_target_root_asset_partitions(self) -> Iterable[AssetKeyPartitionKey]:
-        root_asset_keys = (
-            AssetSelection.keys(*self.target_subset.asset_keys)
-            .sources()
-            .resolve(self.target_subset.asset_graph)
-        )
-        return list(
-            self.target_subset.filter_asset_keys(root_asset_keys).iterate_asset_partitions()
-        )
+        connected_asset_sets = []
+        all_visited_asset_keys = set()
```

Comment:
@sryza I believe it's O(N) -- we visit each targeted asset once and find all connected targeted neighbors, caching the 'visited' neighbors in `all_visited_asset_keys` so we don't visit again

Reply:
Ahh I see. Yes you are right, this is O(N).

That said, it still might be simpler to just look for assets in the target subset with no parents in the target subset?

Reply:
good call -- agree that seems much simpler